File: ios/joinery-kit/Sources/JoineryMailKit/MailboxStore.swift
import Foundation
import JoineryKit

/// State for the mailbox screen: the granted mailboxes plus the thread list
/// for the current mailbox / view / search, with paging. All mutations go
/// through MailAPI and re-read or locally patch the list — the server is the
/// single source of truth shared with the web reader.
@MainActor
public final class MailboxStore: ObservableObject {
    public enum Phase {
        case loading
        case loaded
        case failed(String)
    }

    @Published public private(set) var phase: Phase = .loading
    @Published public private(set) var home: MailboxHome?
    @Published public private(set) var threads: [ThreadSummary] = []
    @Published public private(set) var hasMore = false
    @Published public private(set) var isLoadingMore = false

    @Published public var searchText = ""
    @Published public private(set) var activeQuery = ""
    @Published public private(set) var view: MailView = .inbox
    @Published public private(set) var selectedAlias: Int?

    public let api: MailAPI
    private var page = 1
    /// Ignores stale in-flight loads after the view/mailbox/search changes.
    private var loadGeneration = 0

    public init(api: MailAPI) {
        self.api = api
    }

    /// The mailbox the list is scoped to, when a specific one is selected.
    public var selectedMailbox: Mailbox? {
        guard let selectedAlias else { return nil }
        return home?.mailboxes.first { $0.aliasID == selectedAlias }
    }

    public var title: String {
        if activeQuery.isEmpty == false { return "Search" }
        return view.title
    }

    /// First load: mailboxes and the initial thread page together.
    public func initialLoad() async {
        phase = .loading
        do {
            async let homeTask = api.mailboxes()
            async let pageTask = api.threadList(aliasID: selectedAlias, view: view, query: activeQuery, page: 1)
            let (loadedHome, firstPage) = try await (homeTask, pageTask)
            home = loadedHome
            apply(firstPage, reset: true)
            phase = .loaded
        } catch {
            phase = .failed((error as? JoineryAPIError)?.displayMessage ?? error.localizedDescription)
        }
    }

    /// Re-read the current slice from page 1 (pull-to-refresh, after actions,
    /// after a view/mailbox/search change). Keeps showing the last-good list
    /// while it runs; failures surface only when nothing is loaded yet.
    public func reload(refreshMailboxes: Bool = false) async {
        loadGeneration += 1
        let generation = loadGeneration
        do {
            if refreshMailboxes {
                home = try await api.mailboxes()
            }
            let firstPage = try await api.threadList(aliasID: selectedAlias, view: view, query: activeQuery, page: 1)
            guard generation == loadGeneration else { return }
            apply(firstPage, reset: true)
            phase = .loaded
        } catch {
            guard generation == loadGeneration else { return }
            if case .loaded = phase { return }
            phase = .failed((error as? JoineryAPIError)?.displayMessage ?? error.localizedDescription)
        }
    }

    public func loadMore() async {
        guard hasMore, !isLoadingMore else { return }
        isLoadingMore = true
        defer { isLoadingMore = false }
        let generation = loadGeneration
        do {
            let next = try await api.threadList(aliasID: selectedAlias, view: view, query: activeQuery, page: page + 1)
            guard generation == loadGeneration else { return }
            apply(next, reset: false)
        } catch {
            // Paging failures are silent; the next scroll retries.
        }
    }

    private func apply(_ pageData: ThreadPage, reset: Bool) {
        if reset {
            threads = pageData.threads
        } else {
            let known = Set(threads.map(\.threadKey))
            threads += pageData.threads.filter { !known.contains($0.threadKey) }
        }
        page = pageData.page
        hasMore = pageData.hasMore
    }

    // MARK: Slice changes

    public func select(view newView: MailView) async {
        guard newView != view else { return }
        view = newView
        await reload()
    }

    public func select(alias: Int?) async {
        guard alias != selectedAlias else { return }
        selectedAlias = alias
        await reload()
    }

    public func submitSearch() async {
        activeQuery = searchText.trimmingCharacters(in: .whitespaces)
        await reload()
    }

    public func clearSearch() async {
        guard !activeQuery.isEmpty else { return }
        searchText = ""
        activeQuery = ""
        await reload()
    }

    // MARK: Row actions (list swipes; detail actions reload on return)

    /// Run a thread action and patch the row locally so the list responds
    /// instantly; a background reload then reconciles with the server.
    public func perform(_ action: String, on thread: ThreadSummary) async {
        do {
            try await api.threadAction(action, threadKey: thread.threadKey, aliasID: selectedAlias)
        } catch {
            await reload()
            return
        }
        switch action {
        case "mark_read":
            patch(thread.threadKey) { $0.unreadCount = 0 }
        case "mark_unread":
            patch(thread.threadKey) { $0.unreadCount = max(1, $0.unreadCount) }
        case "star":
            patch(thread.threadKey) { $0.isStarred = true }
        case "unstar":
            patch(thread.threadKey) { $0.isStarred = false }
        case "archive":
            if case .inbox = view { remove(thread.threadKey) } else {
                patch(thread.threadKey) { $0.isArchived = true }
            }
        case "unarchive":
            patch(thread.threadKey) { $0.isArchived = false }
        case "delete", "mark_spam", "mark_not_spam":
            remove(thread.threadKey)
        default:
            await reload()
        }
    }

    /// Local patch used by the detail screen when it changes thread state.
    public func patch(_ threadKey: String, mutate: (inout ThreadSummary) -> Void) {
        guard let index = threads.firstIndex(where: { $0.threadKey == threadKey }) else { return }
        mutate(&threads[index])
    }

    public func remove(_ threadKey: String) {
        threads.removeAll { $0.threadKey == threadKey }
    }
}

File: android/joinery-android-mail/src/main/java/com/getjoinery/mail/MailboxStore.kt
package com.getjoinery.mail

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.getjoinery.android.JoineryApiError
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * State for the mailbox screen: the granted mailboxes plus the thread list
 * for the current mailbox / view / folder / search, with paging. All
 * mutations go through [MailApi] and re-read or locally patch the list — the
 * server is the single source of truth shared with the web reader.
 */
class MailboxStore(val api: MailApi) {
    sealed class Phase {
        object Loading : Phase()
        object Loaded : Phase()
        data class Failed(val message: String) : Phase()
    }

    var phase by mutableStateOf<Phase>(Phase.Loading)
        private set
    var home by mutableStateOf<MailboxHome?>(null)
        private set
    var threads by mutableStateOf<List<ThreadSummary>>(emptyList())
        private set
    var hasMore by mutableStateOf(false)
        private set
    var isLoadingMore by mutableStateOf(false)
        private set

    var searchText by mutableStateOf("")
    var activeQuery by mutableStateOf("")
        private set
    var view by mutableStateOf(MailView.INBOX)
        private set
    var selectedAlias by mutableStateOf<Int?>(null)
        private set
    /** A selected folder replaces the view as the list's slice. */
    var selectedFolder by mutableStateOf<MailFolder?>(null)
        private set

    private var page = 1
    /** Ignores stale in-flight loads after the view/mailbox/search changes. */
    private var loadGeneration = 0

    /** The mailbox the list is scoped to, when a specific one is selected. */
    val selectedMailbox: Mailbox?
        get() = selectedAlias?.let { alias -> home?.mailboxes?.firstOrNull { it.aliasId == alias } }

    /** The mailbox whose folder rail applies: the selected one, or the only
     *  grant when there is exactly one (the switcher is hidden then). */
    val effectiveMailbox: Mailbox?
        get() = selectedMailbox ?: home?.mailboxes?.singleOrNull()

    val title: String
        get() = when {
            activeQuery.isNotEmpty() -> "Search"
            selectedFolder != null -> selectedFolder!!.name
            else -> view.title
        }

    /** First load: mailboxes and the initial thread page together. */
    suspend fun initialLoad() {
        phase = Phase.Loading
        try {
            coroutineScope {
                val homeTask = async { api.mailboxes() }
                val pageTask = async {
                    api.threadList(selectedAlias, view, selectedFolder?.id, activeQuery, 1)
                }
                home = homeTask.await()
                apply(pageTask.await(), reset = true)
            }
            phase = Phase.Loaded
        } catch (e: Exception) {
            phase = Phase.Failed(displayMessage(e))
        }
    }

    /** Re-read the current slice from page 1 (pull-to-refresh, after actions,
     *  after a view/mailbox/search change). Keeps showing the last-good list
     *  while it runs; failures surface only when nothing is loaded yet. */
    suspend fun reload(refreshMailboxes: Boolean = false) {
        loadGeneration += 1
        val generation = loadGeneration
        try {
            if (refreshMailboxes) {
                home = api.mailboxes()
            }
            val firstPage = api.threadList(selectedAlias, view, selectedFolder?.id, activeQuery, 1)
            if (generation != loadGeneration) return
            apply(firstPage, reset = true)
            phase = Phase.Loaded
        } catch (e: Exception) {
            if (generation != loadGeneration) return
            if (phase is Phase.Loaded) return
            phase = Phase.Failed(displayMessage(e))
        }
    }

    suspend fun loadMore() {
        if (!hasMore || isLoadingMore) return
        isLoadingMore = true
        val generation = loadGeneration
        try {
            val next = api.threadList(selectedAlias, view, selectedFolder?.id, activeQuery, page + 1)
            if (generation == loadGeneration) apply(next, reset = false)
        } catch (e: Exception) {
            // Paging failures are silent; the next scroll retries.
        } finally {
            isLoadingMore = false
        }
    }

    private fun apply(pageData: ThreadPage, reset: Boolean) {
        threads = if (reset) {
            pageData.threads
        } else {
            val known = threads.map { it.threadKey }.toHashSet()
            threads + pageData.threads.filter { !known.contains(it.threadKey) }
        }
        page = pageData.page
        hasMore = pageData.hasMore
    }

    // MARK: Slice changes

    suspend fun select(newView: MailView) {
        if (newView == view && selectedFolder == null) return
        view = newView
        selectedFolder = null
        reload()
    }

    suspend fun select(alias: Int?) {
        if (alias == selectedAlias) return
        selectedAlias = alias
        selectedFolder = null
        reload()
    }

    /** Scope the list to one folder of one mailbox (the folder rail). */
    suspend fun select(folder: MailFolder, ofAlias: Int) {
        if (selectedFolder?.id == folder.id && selectedAlias == ofAlias) return
        selectedAlias = ofAlias
        selectedFolder = folder
        reload()
    }

    suspend fun submitSearch() {
        activeQuery = searchText.trim()
        reload()
    }

    suspend fun clearSearch() {
        if (activeQuery.isEmpty()) return
        searchText = ""
        activeQuery = ""
        reload()
    }

    // MARK: Row actions (list swipes; detail actions patch through here too)

    /** Run a thread action and patch the row locally so the list responds
     *  instantly; failures fall back to a reload to reconcile. */
    suspend fun perform(action: String, thread: ThreadSummary) {
        try {
            api.threadAction(action, thread.threadKey, selectedAlias)
        } catch (e: Exception) {
            reload()
            return
        }
        when (action) {
            "mark_read" -> patch(thread.threadKey) { it.copy(unreadCount = 0) }
            "mark_unread" -> patch(thread.threadKey) { it.copy(unreadCount = maxOf(1, it.unreadCount)) }
            "star" -> patch(thread.threadKey) { it.copy(isStarred = true) }
            "unstar" -> patch(thread.threadKey) { it.copy(isStarred = false) }
            "archive" ->
                if (view == MailView.INBOX && selectedFolder == null) {
                    remove(thread.threadKey)
                } else {
                    patch(thread.threadKey) { it.copy(isArchived = true) }
                }
            "unarchive" -> patch(thread.threadKey) { it.copy(isArchived = false) }
            "delete", "mark_spam", "mark_not_spam" -> remove(thread.threadKey)
            else -> reload()
        }
    }

    /** Local patch used by the detail screen when it changes thread state. */
    fun patch(threadKey: String, mutate: (ThreadSummary) -> ThreadSummary) {
        threads = threads.map { if (it.threadKey == threadKey) mutate(it) else it }
    }

    fun remove(threadKey: String) {
        threads = threads.filter { it.threadKey != threadKey }
    }

    private fun displayMessage(e: Exception): String =
        (e as? JoineryApiError)?.displayMessage ?: (e.message ?: "Something went wrong.")
}
